src/bulk/common.ts

```diff
@@ -1,43 +1,198 @@
+import type { CollationOptions } from '../cmap/wire_protocol/write_command';
+import type { InsertOptions } from './../operations/insert';
+import type { RemoveOptions } from './../cmap/wire_protocol/index';
+import type { ReplaceOptions } from './../operations/replace_one';
+import type { UpdateOptions } from './../operations/update';
+
+interface FinalOptionsConfigOptions {
+  bypassDocumentValidation?: boolean;
+  retryWrites?: boolean;
+}
+interface ResultHandler {
+  (err?: any, result?: any): void;
+  operationId?: number;
+}
 
 // Error codes
 const WRITE_CONCERN_ERROR = 64;
 
 // Insert types
-const INSERT = 1;
-const UPDATE = 2;
-const REMOVE = 3;
+export const INSERT = 1;
+export const UPDATE = 2;
+export const REMOVE = 3;
+
+/** Number indicating the document type */
+export type BatchTypes = typeof INSERT | typeof UPDATE | typeof REMOVE;
+
+/** @public */
+export interface Operation {
+  selector: Document;
+  multi: boolean;
+  upsert: boolean;
+  limit: number;
+}
+
+/** @public */
+export interface InsertOneModel extends Document, InsertOptions {
+  document?: Document;
+  _id?: ObjectId;
+}
+
+/** @public */
+export type InsertManyModel = InsertOneModel[];
+
+/** @public */
+export interface ReplaceOneModel extends ReplaceOptions {
+  q?: Document;
+  replacement: Document;
+  filter?: Document;
+}
+
+/** @public */
+export interface UpdateOneModel extends UpdateOptions {
+  q?: Document;
+  update: Document;
+  filter?: Document;
+}
+
+/** @public */
+export interface UpdateManyModel extends UpdateOptions {
+  q?: Document;
+  update: Document;
+  filter?: Document;
+}
+
+/** @public */
+export type RemoveOneModel = RemoveOptions;
+
+/** @public */
+export interface RemoveManyModel extends RemoveOptions {
+  limit?: number;
+}
+
+/** @public */
+export interface DeleteOneModel extends RemoveOptions {
+  q?: Document;
+}
+
+/** @public */
+export interface DeleteManyModel extends RemoveOptions {
+  q?: Document;
+}
+
+/** @public */
+export interface BatchOperations {
+  insertOne: InsertOneModel;
+  insertMany: InsertManyModel;
+  replaceOne: ReplaceOneModel;
+  updateOne: UpdateOneModel;
+  updateMany: UpdateManyModel;
+  removeOne: RemoveOneModel;
+  removeMany: RemoveManyModel;
+  deleteOne: DeleteOneModel;
+  deleteMany: DeleteManyModel;
+}
+
+/** @public */
+export type AnyModel = InsertOneModel &
+  InsertManyModel &
+  ReplaceOneModel &
+  UpdateOneModel &
+  UpdateManyModel &
+  RemoveOneModel &
+  RemoveManyModel &
+  DeleteOneModel &
+  DeleteManyModel;
+
+/** @public */
+export type AnyOperation =
+  | { insertOne: InsertOneModel }
+  | { insertMany: InsertManyModel }
+  | { replaceOne: ReplaceOneModel }
+  | { updateOne: UpdateOneModel }
+  | { updateMany: UpdateManyModel }
+  | { removeOne: RemoveOneModel }
+  | { removeMany: RemoveManyModel }
+  | { deleteOne: DeleteOneModel }
+  | { deleteMany: DeleteManyModel };
+
+/** @public */
+export type Operations = AnyOperation[];
```

Comment:
I think we can drop this type in favor of just using `AnyBulkOperation[]` where this is used, I think this type used to be more complex maybe?

Reply:
I'm down for this, will implement direct array use and  wait on consensus for naming.